scripts/build: improve whitespace / empty parameter handling in syscall scripts

lonkaars opened the pull request:
Syscall regex matches may include whitespace characters other than space (i.e. `\n` or `\t`), which causes `gen_syscalls.py` to fail. This mainly affects long system call signatures wrapped onto multiple lines (e.g. by code formatter). This change replaces all successive whitespace in the regex match groups with a single space.

lonkaars committed:
scripts/build: flatten whitespace in `parse_syscalls.py`

Syscall regex matches may include whitespace characters other than space
(i.e. `\n` or `\t`), which causes `gen_syscalls.py` to fail. This mainly
affects long system call signatures wrapped onto multiple lines (e.g. by
code formatter). This change replaces all successive whitespace in the
regex match groups with a single space.

Signed-off-by: Loek Le Blansch <loek.le-blansch.pv@renesas.com>

tejlmand commented:
@lonkaars thanks for this.
Before approving then I would like to know if you have some examples where current parse_syscalls.py fails ?
Preferable in Zephyr repo, but external pointer is also ok.

This can help to see of we can make a test-case / update existing test case to ensure the problem is indeed fixed as well as ensureing there are no side-effects.

lonkaars commented:
I'm currently working in a private repository so I can't share the original code that this fix is for. I've created a twister test case in my (local) Zephyr fork clone with the following function signature, which fails in the same way:
```c
__syscall void
test(int foo);
```

Should I add `// clang-format off`/`on` around the signature or does clang-format not touch tests? The repository I initially encountered this issue on uses `AlignAfterOpenBracket: BlockIndent` which is different from Zephyr's clang-format configuration.

---

While making the test case I also found `gen_syscalls.py` fails when a syscall is declared without parameters. Should I fix and append it to this PR or open a separate one?

tejlmand commented:
> I'm currently working in a private repository so I can't share the original code that this fix is for.

I was assuming so.

> I've created a twister test case in my (local) Zephyr fork clone with the following function signature, which fails in the same way:
> ```
> __syscall void
> test(int foo);
> ```

would be nice to have as part of this PR.
Most tooling is indirectly tested as it's used by Zephyr itself, but when having parts that are not indirectly tested in-tree, such as this line-wrap support, then having test-case to unsure functionality is always good.

> While making the test case I also found gen_syscalls.py fails when a syscall is declared without parameters. Should I fix and append it to this PR or open a separate one?

Making it part of this PR is fine.

lonkaars commented:
Looking over the commit history on [`scripts/build/gen_syscalls.py`](https://github.com/zephyrproject-rtos/zephyr/commits/main/scripts/build/gen_syscalls.py) I believe 1993ea019b15857786b624b889198e32862ada59 can also be reverted as the `\s+` match will cause both newlines and tab characters to be replaced in the json file generated by `parse_syscalls.py`.

tejlmand commented on tests/cmake/syscalls/main.h:
```diff
@@ -0,0 +1,15 @@
+
+#include <zephyr/syscall.h>
+
+// fixed in commit 54e369998747ccb89c2e563c4167b4a90c8670b2
```
be careful with commit references like this because the SHA will change when the PR is merge (rebased) on main.
This means that in the future then this sha can no longer be retrieved.

lonkaars committed:
tests/cmake: add test for syscall generation scripts

This commit adds a test for whitespace in syscall declaration function
signatures.

Signed-off-by: Loek Le Blansch <loek.le-blansch.pv@renesas.com>

lonkaars committed:
revert "build: support newlines in syscall decls"

This reverts commit 1993ea019b15857786b624b889198e32862ada59 because the
`parse_syscalls.py` script has been modified to not output whitespace
characters other than spaces in the generated JSON file.

Signed-off-by: Loek Le Blansch <loek.le-blansch.pv@renesas.com>

tejlmand reviewed (dismissed):
Thanks for the work

dcpleung commented on scripts/build/gen_syscalls.py:
```diff
@@ -411,7 +411,7 @@ def analyze_fn(match_group, fn, userspace_only):
 
     try:
-        if args == "void":
+        if args == "void" or len(args) == 0:
```
IIRC, compliance check would complain about empty argument list that has no `void` in it. If that's the case, I don't think we need this here.
